rholang/src/main/scala/coop/rchain/rholang/interpreter/match.scala

```diff
@@ -0,0 +1,273 @@
+package coop.rchain.rholang.interpreter
+
+import cats._
+import cats.data._
+import cats.implicits._
+
+import scala.annotation.tailrec
+import scala.collection.immutable.Stream
+
+object SpatialMatcher {
+  type FreeMap            = Map[Int, Par]
+  type OptionalFreeMap[A] = StateT[Option, FreeMap, A]
+  type NonDetFreeMap[A]   = StateT[Stream, FreeMap, A]
+
+  def possiblyRemove[T](needle: T, haystack: List[T]): Option[List[T]] = {
+    val (before, after) = haystack.span(x => x != needle)
+    after match {
+      case Nil       => None
+      case _ :: tail => Some(before ++ tail)
+    }
+  }
+
+  @tailrec
+  def possiblyFind[T, R](prop: (T => Option[R]), haystack: List[T]): Option[R] =
+    haystack match {
+      case Nil => None
+      case head :: rest =>
+        prop(head) match {
+          case None    => possiblyFind(prop, rest)
+          case success => success
+        }
+    }
+
+  @tailrec
+  def exists[T](prop: (T => Boolean), haystack: List[T]): Boolean =
+    haystack match {
+      case Nil => false
+      case head :: rest =>
+        if (prop(head))
```

Comment:
God.. `cats` are so limited. There is a `fold` on `BooleanOps` at scalaz, but not in cats. We have our own `catscontrib.BooleanOps` (still in comm, I have to extract that as a library)

Reply:
Take a look at [mouse project](https://github.com/typelevel/mouse)